drivers/watchdog/bflb_wdt.c

```diff
@@ -0,0 +1,297 @@
+{
+	unsigned int timeout_ticks;
+	struct bflb_watchdog_device *bflb_wdd = to_bflb_wdd(wdd);
+
+	bflb_unlock_watchdog(bflb_wdd);
+	timeout_ticks = wdd->timeout * BFLB_TICKS_PER_SEC;
+	writew(timeout_ticks, bflb_wdd->regs + BFLB_REG_WMR);
+}
+
+static int bflb_wdt_set_timeout(struct watchdog_device *wdd,
+				unsigned int timeout)
+{
+	if (timeout >= wdd->max_timeout) {
+		dev_warn(wdd->parent,
+			"timeout %i > max_timeout %i, using max_timeout...",
+			timeout, wdd->max_timeout);
+		timeout = wdd->max_timeout;
+	}
+
+	wdd->timeout = timeout;
+
+	bflb_wdt_update_timeout_reg(wdd);
+
+	dev_dbg(wdd->parent, "bflb_wdt_set_timeout (s=%i tps=%i)",
+		timeout, BFLB_TICKS_PER_SEC);
+
+	return 0;
+}
+
+static int bflb_wdt_start(struct watchdog_device *wdd)
+{
+	uint32_t reg_val;
+	struct bflb_watchdog_device *bflb_wdd = to_bflb_wdd(wdd);
+
+
+	// And enable the watchdog
+	bflb_unlock_watchdog(bflb_wdd);
+	reg_val = readl(bflb_wdd->regs + BFLB_REG_WMER);
+	reg_val |= BFLB_VAL_WE;
+	writel(reg_val, bflb_wdd->regs + BFLB_REG_WMER);
+
+	dev_info(wdd->parent, "bflb_wdt_start started...");
+
+	return 0;
+}
+
+static int bflb_wdt_stop(struct watchdog_device *wdd)
+{
+	uint32_t reg_val;
+	struct bflb_watchdog_device *bflb_wdd = to_bflb_wdd(wdd);
+
+	// disable
+	bflb_unlock_watchdog(bflb_wdd);
+	reg_val = readl(bflb_wdd->regs + BFLB_REG_WMER);
+	reg_val &= ~BFLB_VAL_WE;
+	writel(reg_val, bflb_wdd->regs + BFLB_REG_WMER);
+
+	dev_info(wdd->parent, "bflb_wdt_stopped...");
+
+	return 0;
+};
+
+
+static unsigned int bflb_wdt_timeleft(struct watchdog_device *wdd)
+{
+	unsigned int used_seconds;
+	unsigned int remaining_seconds;
+	unsigned int ticks;
+	struct bflb_watchdog_device *bflb_wdd = to_bflb_wdd(wdd);
+
+	ticks = readw(bflb_wdd->regs + BFLB_REG_WVR);
+
+	used_seconds = ticks / BFLB_TICKS_PER_SEC;
+	remaining_seconds = wdd->max_timeout - used_seconds;
+	dev_dbg(wdd->parent, "bflb_wdt_time left %i (elapsed tick %i, sec %i)",
+		remaining_seconds, ticks, used_seconds);
+
+	return remaining_seconds;
+};
+
+static const struct watchdog_info bflb_wdt_info = {
+	.identity = "bflb_wdt",
+	.options = WDIOF_SETTIMEOUT |
+	WDIOF_KEEPALIVEPING |
+	WDIOF_MAGICCLOSE,
+};
+
+static const struct watchdog_ops bflb_wdt_ops = {
+	.start = bflb_wdt_start,
+	.stop = bflb_wdt_stop,
+	.ping = bflb_wdt_ping,
+	.set_timeout = bflb_wdt_set_timeout,
+	.get_timeleft = bflb_wdt_timeleft,
+};
+
+
+static int __init bflb_wdt_probe(struct platform_device *pdev)
+{
+	struct bflb_watchdog_device *bflb_wdd;
+	struct watchdog_device *wdd;
+	struct resource *res;
+	int err;
+	uint32_t reg_val;
+
+	dev_dbg(&pdev->dev, "bflb_wdt_probe started");
+
+	bflb_wdd = devm_kzalloc(&pdev->dev, sizeof(*bflb_wdd), GFP_KERNEL);
+	if (!bflb_wdd) return -ENOMEM;
+
+	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	bflb_wdd->regs = devm_ioremap_resource(&pdev->dev, res);
+	if (IS_ERR(bflb_wdd->regs))
+		return PTR_ERR(bflb_wdd->regs);
```

Comment:
Please use `devm_platform_ioremap_resource`. If you don't switch to it, someone else will :)

Reply:
Fixed in [a8f44c8](https://github.com/arm000/linux-bl808/pull/8/commits/a8f44c89b43f6246175c6044d3184264cb493dd2)